File: packages/junoguard/src/render.ts
/**
 * JunoGuard render layer — one visual language for every surface.
 *
 * Emits lines of styled spans rather than finished strings, so the MCP server
 * can flatten them to plain text and the CLI can paint them, without the two
 * drifting apart. Weight matches severity: an allow is one quiet line because
 * it happens constantly, a flag is a square-bordered caution, a block is a
 * rounded-border refusal with the blast radius under it.
 */

import type { BlastRadius, InstallResult, LlmResult, StatusResult, Verdict } from "./types.js";

export type Style = "" | "block" | "flag" | "allow" | "label" | "dim" | "key";
export type Span = [text: string, style: Style];
export type Line = Span[];

const BOX_W = 54; // inner width, between the border characters
const LABEL_W = 11; // "VERDICT    "
const SUB_W = 23; // "credentials in scope   "

// --------------------------------------------------------------------------
// primitives
// --------------------------------------------------------------------------

export function toPlain(lines: Line[]): string {
  return lines.map((line) => line.map(([text]) => text).join("")).join("\n");
}

const blank = (): Line => [];
const text = (value: string, style: Style = ""): Line => [[value, style]];
const pad = (value: string, width: number) => value.padEnd(width);

/**
 * The heavy header used by flag and block. Rounded corners read as more severe
 * than square ones, so block gets the rounded box and flag the square one.
 */
function banner(title: string, subject: string, style: Style, rounded: boolean): Line[] {
  const [tl, tr, bl, br] = rounded ? ["╭", "╮", "╰", "╯"] : ["┌", "┐", "└", "┘"];
  const cap = `─ ${title} `;
  const fill = "─".repeat(Math.max(0, BOX_W - cap.length));
  const trimmed = subject.slice(0, BOX_W - 4);
  return [
    [[`${tl}${cap}${fill}${tr}`, style]],
    [
      ["│", style],
      [pad(`  ${trimmed}`, BOX_W), ""],
      ["│", style],
    ],
    [[`${bl}${"─".repeat(BOX_W)}${br}`, style]],
  ];
}

const field = (label: string, value: string, style: Style = ""): Line => [
  [pad(label, LABEL_W), "label"],
  [value, style],
];

/** A continuation line under a field label. */
const cont = (value: string, style: Style = ""): Line => [
  [" ".repeat(LABEL_W), ""],
  [value, style],
];

const sub = (label: string, value: string, style: Style = ""): Line => [
  [" ".repeat(LABEL_W), ""],
  [pad(label, SUB_W), "dim"],
  [value, style],
];

/** Wrap a comma-separated list to `width`, keeping the commas trailing. */
function wrap(items: string[], width: number): string[] {
  const out: string[] = [];
  let current = "";
  items.forEach((item, index) => {
    const piece = item + (index < items.length - 1 ? "," : "");
    if (current && current.length + 1 + piece.length > width) {
      out.push(current);
      current = piece;
    } else {
      current = current ? `${current} ${piece}` : piece;
    }
  });
  if (current) out.push(current);
  return out.length ? out : [""];
}

/** Two decimals for round figures like a budget, four for fractional spend. */
function money(value: number | null | undefined): string {
  if (value === null || value === undefined || Number.isNaN(value)) return "n/a";
  const places = Number(value.toFixed(2)) === Number(value.toFixed(6)) ? 2 : 4;
  return `$${value.toLocaleString("en-US", {
    minimumFractionDigits: places,
    maximumFractionDigits: places,
  })}`;
}

const count = (value: number) => value.toLocaleString("en-US");

/** How the verdict was reached, phrased for a human. */
function sourceOf(verdict: Verdict | undefined): string {
  const source = verdict?.source ?? "unknown";
  const known: Record<string, string> = {
    ossprey: "via Ossprey",
    cache: "via Ossprey (cached)",
    mock: "via Ossprey (offline fixture)",
  };
  return known[source] ?? `via ${source}`;
}

function bar(fraction: number, width = 20): string {
  const clamped = Math.min(Math.max(fraction, 0), 1);
  const filled = Math.round(clamped * width);
  return "█".repeat(filled) + "░".repeat(width - filled);
}

// --------------------------------------------------------------------------
// blast radius
// --------------------------------------------------------------------------

function blastRadius(blast: BlastRadius | null | undefined, preamble: string): Line[] {
  if (!blast) return [];
  const lines: Line[] = [
    blank(),
    [
      ["BLAST RADIUS  ", "label"],
      [preamble, "dim"],
    ],
  ];

  const creds = blast.credentials_in_scope ?? [];
  if (creds.length) {
    const wrapped = wrap(creds, BOX_W - SUB_W);
    lines.push(sub("credentials in scope", wrapped[0]!, "key"));
    for (const extra of wrapped.slice(1)) {
      lines.push([
        [" ".repeat(LABEL_W + SUB_W), ""],
        [extra, "key"],
      ]);
    }
  }

  if (blast.network_egress) lines.push(sub("network egress", String(blast.network_egress)));
  if (blast.write_access) lines.push(sub("write access", String(blast.write_access)));

  if (blast.summary) {
    lines.push(blank());
    lines.push(cont(`→ ${blast.summary}`, "block"));
  }
  return lines;
}

// --------------------------------------------------------------------------
// Lane A — package installs
// --------------------------------------------------------------------------

export function renderInstall(payload: InstallResult, pkg: string, ecosystem: string): Line[] {
  const decision = payload.decision ?? "flag";
  const verdict = payload.verdict ?? {};
  const findings = verdict.findings ?? [];
  const severity = verdict.severity ?? payload.risk_level ?? "unknown";

  if (decision === "allow") {
    // Allows happen constantly. One quiet line, nothing more.
    const detail = severity === "unknown" ? "no findings" : severity;
    return [
      [
        ["✓ juno ", "allow"],
        [`· allow — ${pkg} (${ecosystem}) · ${detail}`, "dim"],
      ],
    ];
  }

  const blocked = decision === "block";
  const style: Style = blocked ? "block" : "flag";
  // "We could not look" is a different message from "we found something": the
  // agent has to be told to retry, not to go and pick another dependency.
  const unscanned = payload.review_required === true || verdict.available === false;
  const title = blocked
    ? unscanned
      ? "JUNO · NOT SCANNED"
      : "JUNO · BLOCKED"
    : "JUNO · FLAGGED";

  const lines = banner(title, `${pkg}  (${ecosystem})`, style, blocked);
  lines.push(blank());
  lines.push(field("VERDICT", `${severity}  ·  ${sourceOf(verdict)}`, style));
  for (const finding of findings) lines.push(cont(String(finding)));
  if (!findings.length && payload.reason) lines.push(cont(String(payload.reason)));

  lines.push(
    ...blastRadius(
      payload.blast_radius,
      blocked ? "if this had installed" : "if this turns out to be hostile",
    ),
  );

  lines.push(blank());
  if (blocked && unscanned) {
    const retry = payload.retry_after_seconds;
    lines.push(text("This package was not installed, and it was never scanned.", style));
    lines.push(
      text(
        retry
          ? `The scanner is unavailable — retry in ${retry}s, or get an operator review.`
          : "The scanner is unavailable — retry later, or get an operator review.",
        style,
      ),
    );
    lines.push(text("This is not a finding against the package.", "dim"));
  } else if (blocked) {
    // The agent reads this. It has to be unambiguous enough that it picks a
    // different dependency instead of retrying the same one.
    lines.push(text("This package was not installed. Choose a different dependency.", style));
  } else {
    lines.push(text("Not blocked, but unverified. Prefer a dependency with a known", "flag"));
    lines.push(text("publisher if one exists; if you proceed, say why.", "flag"));
  }
  return lines;
}

// --------------------------------------------------------------------------
// Lane B — model calls
// --------------------------------------------------------------------------

function budgetLines(payload: LlmResult): Line[] {
  const spend = payload.spend_today_usd;
  const budget = payload.daily_budget_usd;
  if (spend === undefined || !budget) return [];
  return [
    sub("spend today", `${money(spend)} of ${money(budget)}`),
    sub("remaining", money(Math.max(0, budget - spend))),
  ];
}

export function renderLlm(payload: LlmResult, model: string): Line[] {
  const decision = payload.decision ?? "flag";
  const spend = payload.spend_today_usd;
  const budget = payload.daily_budget_usd;

  if (decision === "allow") {
    const bits = [`· allow — ${model}`];
    if (payload.tokens_in !== undefined) {
      bits.push(`${count(payload.tokens_in)} in / ${count(payload.tokens_out ?? 0)} out`);
    }
    if (payload.cost_usd !== undefined) bits.push(`$${payload.cost_usd.toFixed(6)}`);
    if (spend !== undefined && budget) bits.push(`${money(spend)} of ${money(budget)} today`);

    const lines: Line[] = [
      [
        ["✓ juno ", "allow"],
        [bits.join("  ·  "), "dim"],
      ],
    ];
    if (payload.answer) lines.push(blank(), text(String(payload.answer)));
    return lines;
  }

  const blocked = decision === "block";
  const style: Style = blocked ? "block" : "flag";
  const title = blocked ? "JUNO · BLOCKED" : "JUNO · FLAGGED";

  const lines = banner(title, `model call  ·  ${model}`, style, blocked);
  lines.push(blank());
  lines.push(field("REASON", String(payload.reason ?? "policy"), style));
  lines.push(cont(`risk: ${payload.risk_level ?? "unknown"}`, "dim"));

  const budgets = budgetLines(payload);
  if (budgets.length) {
    lines.push(blank(), [["BUDGET", "label"]], ...budgets);
  }

  lines.push(blank());
  if (blocked) {
    lines.push(text("No model call was made and nothing was charged.", style));
    lines.push(text("Do not retry this call. Ask the operator to raise the budget,", style));
    lines.push(text("reduce the request, or stop here.", style));
  } else {
    lines.push(text("Allowed, but this call is outside the normal pattern for this", "flag"));
    lines.push(text("project. Keep the next few calls small.", "flag"));
    if (payload.answer) lines.push(blank(), text(String(payload.answer)));
  }
  return lines;
}

// --------------------------------------------------------------------------
// status
// --------------------------------------------------------------------------

export function renderStatus(payload: StatusResult): Line[] {
  const project = payload.project ?? "project";
  const state = payload.status ?? "unknown";
  const suspended = state !== "active";

  const spend = payload.spend_today_usd ?? 0;
  const budget = payload.daily_budget_usd ?? 0;
  const remaining = payload.remaining_usd ?? Math.max(0, budget - spend);
  const used = budget ? spend / budget : 0;

  const style: Style = suspended ? "block" : "allow";
  const title = suspended ? "JUNO · SUSPENDED" : "JUNO · status";
  const lines = banner(title, `${project}  ·  ${state}`, style, suspended);

  lines.push(blank());
  lines.push(field("BUDGET", `${money(spend)} spent of ${money(budget)}`));
  const usedStyle: Style = used >= 0.9 ? "block" : used >= 0.7 ? "flag" : "allow";
  lines.push([
    [" ".repeat(LABEL_W), ""],
    [bar(used), usedStyle],
    [`  ${Math.round(used * 100)}%`, "dim"],
  ]);
  lines.push(cont(`${money(remaining)} remaining`, "dim"));

  const rpm = payload.requests_last_min;
  const cap = payload.max_requests_per_min;
  if (rpm !== undefined) {
    const over = cap !== undefined && rpm > cap;
    lines.push(blank());
    lines.push(
      field(
        "RATE",
        `${rpm} request${rpm === 1 ? "" : "s"} in the last minute` +
          (cap ? `  (limit ${cap}/min)` : ""),
        over ? "block" : "",
      ),
    );
    if (over) lines.push(cont("burst above policy — requests are being rejected", "block"));
  }

  const blocked = payload.blocked_today;
  const incidents = payload.open_incidents;
  if (blocked !== undefined || incidents !== undefined) {
    const parts: string[] = [];
    if (blocked !== undefined) parts.push(`${blocked} blocked`);
    if (incidents !== undefined) {
      parts.push(`${incidents} open incident${incidents === 1 ? "" : "s"}`);
    }
    lines.push(blank());
    const hot = Boolean(blocked) || Boolean(incidents);
    lines.push(field("TODAY", parts.join("  ·  "), hot ? "flag" : "dim"));
  }

  if (suspended) {
    lines.push(blank());
    lines.push(text("This project is suspended. Both lanes are dark — every install", "block"));
    lines.push(text("and model call will be blocked until an operator resumes it.", "block"));
  }
  return lines;
}

// --------------------------------------------------------------------------
// failure states
// --------------------------------------------------------------------------

/**
 * Rendered when the guard itself could not be consulted. Deliberately shaped
 * like a block: an unreachable guard means the action was not approved, and
 * nothing downstream should read it as permission.
 */
export function renderError(subject: string, detail: string, consequence: string): Line[] {
  const lines = banner("JUNO · UNAVAILABLE", subject, "block", true);
  lines.push(blank());
  lines.push(field("REASON", detail, "block"));
  lines.push(blank());
  lines.push(text(consequence, "block"));
  return lines;
}

/**
 * Shown when no project key is set. Also a refusal — nothing was checked, so
 * nothing is approved — but it fails toward teaching rather than alarm, since
 * the cause is setup rather than an attack.
 */
export function renderNotConfigured(subject: string, apiUrl: string): Line[] {
  const lines = banner("JUNO · NOT CONFIGURED", subject, "flag", false);
  lines.push(blank());
  lines.push(field("MISSING", "JUNO_PROJECT_KEY", "flag"));
  lines.push(cont("Nothing was checked, so nothing is approved.", "dim"));
  lines.push(blank());
  lines.push([["TRY IT", "label"]]);
  lines.push(sub("offline, no gateway", "JUNO_MOCK=1", "key"));
  lines.push(blank());
  lines.push([["CONNECT", "label"]]);
  lines.push(sub("your gateway", `JUNO_API_URL=${apiUrl}`, "key"));
  lines.push(sub("your project key", "JUNO_PROJECT_KEY=…", "key"));
  lines.push(blank());
  lines.push(cont("JunoGuard ships no default key on purpose — a shipped", "dim"));
  lines.push(cont("credential would point you at someone else's gateway.", "dim"));
  return lines;
}

File: cli/juno_cli/render.py
"""JunoGuard render layer — one visual language for every client.

Emits lines of styled spans rather than finished strings, so the MCP server can
flatten them to plain text and the CLI can paint them with rich without the two
surfaces drifting apart.

This file is duplicated verbatim at mcp/juno_mcp/render.py so that each package
installs standalone. Edit both, or copy one over the other.
"""

from __future__ import annotations

# A span is (text, style). Style names are rich style strings; the plain-text
# renderer ignores them entirely.
Span = tuple[str, str]
Line = list[Span]

BOX_W = 54  # inner width, between the border characters
LABEL_W = 11  # "VERDICT    "
SUB_W = 23  # "credentials in scope   "

S_BLOCK = "bold red"
S_FLAG = "bold yellow"
S_ALLOW = "green"
S_LABEL = "bold"
S_DIM = "dim"
S_KEY = "cyan"
S_NONE = ""


# --------------------------------------------------------------------------
# primitives
# --------------------------------------------------------------------------


def to_plain(lines: list[Line]) -> str:
    """Flatten styled lines into plain text."""
    return "\n".join("".join(text for text, _ in line) for line in lines)


def _blank() -> Line:
    return []


def _text(text: str, style: str = S_NONE) -> Line:
    return [(text, style)]


def _banner(title: str, subject: str, style: str, rounded: bool) -> list[Line]:
    """The heavy header used by flag and block states.

    Rounded corners read as more severe than square ones, so block gets the
    rounded box and flag the square one.
    """
    tl, tr, bl, br = ("╭", "╮", "╰", "╯") if rounded else ("┌", "┐", "└", "┘")
    cap = f"─ {title} "
    fill = "─" * max(0, BOX_W - len(cap))
    subject = subject[: BOX_W - 4]
    return [
        [(f"{tl}{cap}{fill}{tr}", style)],
        [("│", style), (f"  {subject}".ljust(BOX_W), S_NONE), ("│", style)],
        [(f"{bl}{'─' * BOX_W}{br}", style)],
    ]


def _field(label: str, value: str, style: str = S_NONE) -> Line:
    return [(label.ljust(LABEL_W), S_LABEL), (value, style)]


def _cont(value: str, style: str = S_NONE) -> Line:
    """A continuation line under a field label."""
    return [(" " * LABEL_W, S_NONE), (value, style)]


def _sub(label: str, value: str, style: str = S_NONE) -> Line:
    return [
        (" " * LABEL_W, S_NONE),
        (label.ljust(SUB_W), S_DIM),
        (value, style),
    ]


def _wrap(items: list[str], width: int) -> list[str]:
    """Wrap a comma-separated list to `width`, keeping the commas trailing."""
    out: list[str] = []
    cur = ""
    for i, item in enumerate(items):
        piece = item + ("," if i < len(items) - 1 else "")
        if cur and len(cur) + 1 + len(piece) > width:
            out.append(cur)
            cur = piece
        else:
            cur = f"{cur} {piece}".strip()
    if cur:
        out.append(cur)
    return out or [""]


def _money(value: float | None) -> str:
    """Two decimals for round figures like a budget, four for fractional spend."""
    if value is None:
        return "n/a"
    if round(value, 2) == round(value, 6):
        return f"${value:,.2f}"
    return f"${value:,.4f}"


def _source(verdict: dict) -> str:
    """How the verdict was reached, phrased for a human."""
    source = (verdict or {}).get("source") or "unknown"
    return {
        "ossprey": "via Ossprey",
        "cache": "via Ossprey (cached)",
        "mock": "via Ossprey (offline fixture)",
    }.get(source, f"via {source}")


def _bar(fraction: float, width: int = 20) -> str:
    fraction = min(max(fraction, 0.0), 1.0)
    filled = int(round(fraction * width))
    return "█" * filled + "░" * (width - filled)


# --------------------------------------------------------------------------
# blast radius
# --------------------------------------------------------------------------


def _blast_radius(blast: dict | None, preamble: str) -> list[Line]:
    if not blast:
        return []
    lines: list[Line] = [_blank(), [("BLAST RADIUS  ", S_LABEL), (preamble, S_DIM)]]

    creds = blast.get("credentials_in_scope") or []
    if creds:
        wrapped = _wrap(list(creds), BOX_W - SUB_W)
        lines.append(_sub("credentials in scope", wrapped[0], S_KEY))
        for extra in wrapped[1:]:
            lines.append([(" " * (LABEL_W + SUB_W), S_NONE), (extra, S_KEY)])

    for label, key in (("network egress", "network_egress"), ("write access", "write_access")):
        if blast.get(key):
            lines.append(_sub(label, str(blast[key])))

    if blast.get("summary"):
        lines.append(_blank())
        lines.append(_cont(f"→ {blast['summary']}", S_BLOCK))
    return lines


# --------------------------------------------------------------------------
# Lane A — package installs
# --------------------------------------------------------------------------


def render_install(payload: dict, package: str, ecosystem: str) -> list[Line]:
    decision = payload.get("decision", "flag")
    verdict = payload.get("verdict") or {}
    findings = verdict.get("findings") or []
    severity = verdict.get("severity") or payload.get("risk_level") or "unknown"
    subject = f"{package}  ({ecosystem})"

    if decision == "allow":
        # Allows happen constantly. One quiet line, nothing more.
        detail = severity if severity != "unknown" else "no findings"
        return [
            [
                ("✓ juno ", S_ALLOW),
                (f"· allow — {package} ({ecosystem}) · {detail}", S_DIM),
            ]
        ]

    blocked = decision == "block"
    style = S_BLOCK if blocked else S_FLAG
    title = "JUNO · BLOCKED" if blocked else "JUNO · FLAGGED"

    lines = _banner(title, subject, style, rounded=blocked)
    lines.append(_blank())
    lines.append(_field("VERDICT", f"{severity}  ·  {_source(verdict)}", style))
    for finding in findings:
        lines.append(_cont(str(finding)))
    if not findings and payload.get("reason"):
        lines.append(_cont(str(payload["reason"])))

    lines += _blast_radius(
        payload.get("blast_radius"),
        "if this had installed" if blocked else "if this turns out to be hostile",
    )

    lines.append(_blank())
    if blocked:
        # The agent reads this. It has to be unambiguous enough that it picks a
        # different dependency instead of retrying the same one.
        lines.append(_text("This package was not installed. Choose a different dependency.", style))
    else:
        lines.append(_text("Not blocked, but unverified. Prefer a dependency with a known", S_FLAG))
        lines.append(_text("publisher if one exists; if you proceed, say why.", S_FLAG))
    return lines


# --------------------------------------------------------------------------
# Lane B — model calls
# --------------------------------------------------------------------------


def _budget_lines(payload: dict) -> list[Line]:
    spend = payload.get("spend_today_usd")
    budget = payload.get("daily_budget_usd")
    if spend is None or not budget:
        return []
    return [
        _sub("spend today", f"{_money(spend)} of {_money(budget)}"),
        _sub("remaining", _money(max(0.0, budget - spend))),
    ]


def render_llm(payload: dict, model: str) -> list[Line]:
    decision = payload.get("decision", "flag")
    spend = payload.get("spend_today_usd")
    budget = payload.get("daily_budget_usd")

    if decision == "allow":
        bits = [f"· allow — {model}"]
        if payload.get("tokens_in") is not None:
            bits.append(f"{payload['tokens_in']:,} in / {payload.get('tokens_out', 0):,} out")
        if payload.get("cost_usd") is not None:
            bits.append(f"${payload['cost_usd']:.6f}")
        if spend is not None and budget:
            bits.append(f"{_money(spend)} of {_money(budget)} today")
        head: list[Line] = [[("✓ juno ", S_ALLOW), ("  ·  ".join(bits), S_DIM)]]
        answer = payload.get("answer")
        if answer:
            head += [_blank(), _text(str(answer))]
        return head

    blocked = decision == "block"
    style = S_BLOCK if blocked else S_FLAG
    title = "JUNO · BLOCKED" if blocked else "JUNO · FLAGGED"

    lines = _banner(title, f"model call  ·  {model}", style, rounded=blocked)
    lines.append(_blank())
    lines.append(_field("REASON", str(payload.get("reason") or "policy"), style))
    lines.append(_cont(f"risk: {payload.get('risk_level', 'unknown')}", S_DIM))

    budget_lines = _budget_lines(payload)
    if budget_lines:
        lines.append(_blank())
        lines.append([("BUDGET", S_LABEL)])
        lines += budget_lines

    lines.append(_blank())
    if blocked:
        lines.append(_text("No model call was made and nothing was charged.", style))
        lines.append(_text("Do not retry this call. Ask the operator to raise the budget,", style))
        lines.append(_text("reduce the request, or stop here.", style))
    else:
        answer = payload.get("answer")
        lines.append(_text("Allowed, but this call is outside the normal pattern for this", S_FLAG))
        lines.append(_text("project. Keep the next few calls small.", S_FLAG))
        if answer:
            lines += [_blank(), _text(str(answer))]
    return lines


# --------------------------------------------------------------------------
# status
# --------------------------------------------------------------------------


def render_status(payload: dict) -> list[Line]:
    project = payload.get("project", "project")
    status = payload.get("status", "unknown")
    suspended = status != "active"

    spend = payload.get("spend_today_usd") or 0.0
    budget = payload.get("daily_budget_usd") or 0.0
    remaining = payload.get("remaining_usd")
    if remaining is None:
        remaining = max(0.0, budget - spend)
    used = (spend / budget) if budget else 0.0

    style = S_BLOCK if suspended else S_ALLOW
    title = "JUNO · SUSPENDED" if suspended else "JUNO · status"
    lines = _banner(title, f"{project}  ·  {status}", style, rounded=suspended)

    lines.append(_blank())
    lines.append(_field("BUDGET", f"{_money(spend)} spent of {_money(budget)}"))
    pct_style = S_BLOCK if used >= 0.9 else S_FLAG if used >= 0.7 else S_ALLOW
    lines.append([(" " * LABEL_W, S_NONE), (_bar(used), pct_style), (f"  {used * 100:.0f}%", S_DIM)])
    lines.append(_cont(f"{_money(remaining)} remaining", S_DIM))

    rpm = payload.get("requests_last_min")
    cap = payload.get("max_requests_per_min")
    if rpm is not None:
        over = cap is not None and rpm > cap
        lines.append(_blank())
        lines.append(
            _field(
                "RATE",
                f"{rpm} request{'s' if rpm != 1 else ''} in the last minute"
                + (f"  (limit {cap}/min)" if cap else ""),
                S_BLOCK if over else S_NONE,
            )
        )
        if over:
            lines.append(_cont("burst above policy — requests are being rejected", S_BLOCK))

    blocked = payload.get("blocked_today")
    incidents = payload.get("open_incidents")
    if blocked is not None or incidents is not None:
        parts = []
        if blocked is not None:
            parts.append(f"{blocked} blocked")
        if incidents is not None:
            parts.append(f"{incidents} open incident{'s' if incidents != 1 else ''}")
        lines.append(_blank())
        hot = bool(blocked) or bool(incidents)
        lines.append(_field("TODAY", "  ·  ".join(parts), S_FLAG if hot else S_DIM))

    if suspended:
        lines.append(_blank())
        lines.append(_text("This project is suspended. Both lanes are dark — every install", S_BLOCK))
        lines.append(_text("and model call will be blocked until an operator resumes it.", S_BLOCK))
    return lines


# --------------------------------------------------------------------------
# failure states
# --------------------------------------------------------------------------


def render_error(subject: str, detail: str, consequence: str) -> list[Line]:
    """Rendered when the guard itself could not be consulted.

    Deliberately shaped like a block: an unreachable guard means the action was
    not approved, and nothing downstream should treat that as permission.
    """
    lines = _banner("JUNO · UNAVAILABLE", subject, S_BLOCK, rounded=True)
    lines.append(_blank())
    lines.append(_field("REASON", detail, S_BLOCK))
    lines.append(_blank())
    lines.append(_text(consequence, S_BLOCK))
    return lines
